_posts/blog/2016-10-28-durability-performance.md

```diff
@@ -0,0 +1,136 @@
+each datanode. One thing I am not sure about, is if these sync operations occur
+in parallel on the replicas on different datanodes.  Lets hope they occur in
+parallel.  The following pointers show where sync occurs in the datanode code.
+
+ * [BlockReceiver.flushOrSync()][fos] calls [ReplicaOutputStreams.syncDataOut()][ros1] and [ReplicaOutputStreams.syncChecksumOut()][ros2] when `isSync` is true.
+ * The methods in ReplicaOutputStreams call [FileChannel.force(true)][fcf] which
+   synchronously flushes data and filesystem metadata.
+
+If files were preallocated (this would avoid syncing local filesystem metadata)
+and checksums were stored in-line, then 1 sync could be done instead of 4.  
+
+## Configuring WAL flush/sync in Accumulo 1.6
+
+Accumulo 1.6.0 only supported `hsync` and this caused [performance
+problems][160_RN_WAL].  In order to offer better performance, the option to
+configure `hflush` was [added in 1.6.1][161_RN_WAL].  The
+[tserver.wal.sync.method][16_UM_SM] configuration option was added to support
+this feature.  This was a tablet server wide option that applied to everything
+written to any table.   
+
+## Group Commit
+
+Each Accumulo Tablet Server has a single WAL.  When multiple clients send
+mutations to a tablet server at around the same time, the tablet sever may group
+all of this into a single WAL operation.  It will do this instead of writing and
+syncing or flushing each client's mutations to the WAL separately.  Doing this
+increase throughput and lowers average latency for clients.
+
+## Configuring WAL flush/sync in Accumulo 1.7+
+
+Accumulo 1.7.0 introduced [table.durability][17_UM_TD], a new per table property
+for configuring durability.  It also stopped using the `tserver.wal.sync.method`
+property.  The `table.durability` property has the following four legal values.
+
+ * **none** : Do not write to WAL            
+ * **log**  : Write to WAL, but do not sync  
+ * **flush** : Write to WAL and call `hflush` 
+ * **sync** : Write to WAL and call `hsync`  
+
+If multiple writes arrive at around the same time with different durability
+settings, then the group commit code will choose the most conservative
+durability.  This can cause one tables settings to slow down writes to another
+table.  
+
+In Accumulo 1.6, it was easy to make all writes use `hflush` because there was
+only one tserver setting.  Getting everything to use `flush` in 1.7 and later
+can be a little tricky because by default the Accumulo metadata table is set to
+use `sync`.  Executing the following command in the Accumulo shell will
+accomplish this (assuming no tables or namespaces have been specifically set to
+`sync`).  The first command sets a system wide table default for `flush`.  The
+second two commands override metadata table specific settings of `sync`.
+
+```
+config -s table.durability=flush
+config -t accumulo.metadata -s table.durability=flush
+config -t accumulo.root -s table.durability=flush
```

Comment:
made updates in f5d8401 ... hopefully this is more clear now... its a very nuanced what I am trying to communicate here... which is why its harder to configure 1.7 correctly for `hflush`